docs/parameters.md
---
layout: default
title: Parameters
nav_order: 8
---

# Robot Parameters
In order to easily change some parameters of the robot's code, they were placed in a common file: [parameters.h](https://github.com/DemonicTricycle/DemonicTricycle-ELECH309/blob/main/src/parameters.h).  
This allowed for testing different values for certain functions quickly between compilations.

## [Code](https://github.com/DemonicTricycle/DemonicTricycle-ELECH309/blob/main/src/parameters.h)

<details open markdown="block">
  <summary>
  parameters.h [Click to open]
  </summary>

```c
// parameters defines

#define SYS_FREQ 80000000L
#define FCY (SYS_FREQ/2)

// for filter.c :
// <editor-fold defaultstate="collapsed" desc="filter.c parameters">

#define FLOORS 4

#define THRESHOLD 230

// filter constants
#define M_FILTER 18 // Q18.14
#define FACTOR_FILTER 262144 // 1 << M
#define FS_FILTER 15000

// </editor-fold>

// for motors.c :
// <editor-fold defaultstate="collapsed" desc="motors.c parameters">

#ifndef WHEELBASE
#define WHEELBASE 0.135 // in m
#endif

#ifndef WHEEL_RADIUS
#define WHEEL_RADIUS 0.04 // in m
#endif

#ifndef NUM_TICKS_PER_TURN
#define NUM_TICKS_PER_TURN 1430 // more or less
#endif

// PWM parameters and macro
#ifndef PWM_TIME_PERIOD
#define PWM_TIME_PERIOD 0.001 // In s, one period is one up AND one down, not just one 'hill' up
#endif
  
#ifndef CLOCK_PRESCALER_BINARY
#define CLOCK_PRESCALER_BINARY 0b00 // In ms
#endif

#ifndef CLOCK_PRESCALER
#define CLOCK_PRESCALER 1 // 0b00 corresponds to a scale of 1
#endif

#define MAX_PWM 0.8 // Max PWM value for the motors

#define QE1_CHANNEL_A_PIN 10 // Assigns channel A of QE1 to RP10(=RB10)
#define QE1_CHANNEL_B_PIN 11 // Assigns channel B of QE1 to RP11(=RB11)
#define QE2_CHANNEL_A_PIN 8  // Assigns channel A of QE2 to RP8(=RB8)
#define QE2_CHANNEL_B_PIN 9  // Assigns channel B of QE2 to RP9(=RB9)
#define CRUISE_ROTATION_SPEED 3.5   // rad / second
#define MAX_ACCELERATION_TIME_ROTATION 1.5 // seconds
#define MAX_ANGLE (CRUISE_ROTATION_SPEED*0.5*(MAX_ACCELERATION_TIME_ROTATION*2))
#define ACCELERATION_ROTATION (CRUISE_ROTATION_SPEED/MAX_ACCELERATION_TIME_ROTATION)
#define ACCEPTABLE_ERROR_TRANSLATION 0.04 // 4cm
#define ACCEPTABLE_ERROR_ROTATION 0.06981 // = (float)4 / 360 * 2 * PI

#define K_TRANSLATION 3.69
#define K_ROTATION (1.03 * 3)
//#define time_interval 3 // ms
  
#define M1_DIR_PIN_AS_INPUT TRISBbits.TRISB2 // set to 0 to output
#define M2_DIR_PIN_AS_INPUT TRISBbits.TRISB3 // set to 0 to output
#define MOTOR_1_DIR_PIN LATBbits.LATB2 // pin used for motor1's direction
#define MOTOR_2_DIR_PIN LATBbits.LATB3 // pin used for motor2's direction

#define ACCELERATION 0.5 // seconds

// </editor-fold>

// <editor-fold defaultstate="collapsed" desc="FSM parameters">
#define FRAME_LENGTH 13 // Frame length in number of bits
#define FRAME_DATA_LENGTH 10 // Frame data length in number of bits
//#define FRAME_PARAMS_LENGTH FRAME_DATA_LENGTH-2 // Frame parameters length in number of bits
// </editor-fold>

```

</details>


docs/microcontrollerConfiguration.md
---
layout: default
title: Microcontroller Configuration
nav_order: 9
---

<details open markdown="block">
  <summary>
    Table of contents
  </summary>
  {: .text-delta }
1. TOC
{:toc}
</details>

# Microcontroller Configuration
The µController used is a [Microchip dsPIC33FJ128MC802](https://www.microchip.com/en-us/product/dsPIC33FJ128MC802).

## Pins Usage
Here is the used pins on the dsPIC µController, and their purpose:  

<img src="assets/images/used_pins.png" height="500" />

For reference, here are the pins of the dsPIC from the datasheet:  

![dsPICpins](assets/images/dsPICpins.png)[^1]

## Configuration Bits
These configuration bits were chosen to allow for the microcontroller's overclock.

The file is available [here](https://github.com/DemonicTricycle/DemonicTricycle-ELECH309/blob/main/src/configuration.c).

<details open markdown="block">
  <summary>
  configuration.c [Click to open]
  </summary>

```c
// DSPIC33FJ128MC802 Configuration Bit Settings
// 'C' source line config statements
// FBS
#pragma config BWRP = WRPROTECT_OFF // Boot Segment Write Protect (Boot Segment may be written)
#pragma config BSS = NO_FLASH       // Boot Segment Program Flash Code Protection (No Boot program Flash segment)
#pragma config RBS = NO_RAM         // Boot Segment RAM Protection (No Boot RAM)

// FSS
#pragma config SWRP = WRPROTECT_OFF // Secure Segment Program Write Protect (Secure segment may be written)
#pragma config SSS = NO_FLASH       // Secure Segment Program Flash Code Protection (No Secure Segment)
#pragma config RSS = NO_RAM         // Secure Segment Data RAM Protection (No Secure RAM)

// FGS
#pragma config GWRP = OFF // General Code Segment Write Protect (User program memory is not write-protected)
#pragma config GSS = OFF  // General Segment Code Protection (User program memory is not code-protected)

// FOSCSEL
#pragma config FNOSC = FRCPLL // Oscillator Mode (Internal Fast RC (FRC))
#pragma config IESO = OFF   // Internal External Switch Over Mode (Start-up device with FRC, then automatically switch to user-selected oscillator source when ready)

// FOSC
#pragma config POSCMD = NONE  // Primary Oscillator Source (Primary Oscillator Disabled)
#pragma config OSCIOFNC = OFF // OSC2 Pin Function (OSC2 pin has clock out function)
#pragma config IOL1WAY = ON   // Peripheral Pin Select Configuration (Allow Only One Re-configuration)
#pragma config FCKSM = CSDCMD // Clock Switching and Monitor (Both Clock Switching and Fail-Safe Clock Monitor are disabled)

// FWDT
#pragma config WDTPOST = PS32768 // Watchdog Timer Postscaler (1:32,768)
#pragma config WDTPRE = PR128    // WDT Prescaler (1:128)
#pragma config WINDIS = OFF      // Watchdog Timer Window (Watchdog Timer in Non-Window mode)
#pragma config FWDTEN = ON      // Watchdog Timer Enable (Watchdog timer enabled/disabled by user software)

// FPOR
#pragma config FPWRT = PWR128 // POR Timer Value (128ms)
#pragma config ALTI2C = OFF   // Alternate I2C  pins (I2C mapped to SDA1/SCL1 pins)
#pragma config LPOL = ON      // Motor Control PWM Low Side Polarity bit (PWM module low side output pins have active-high output polarity)
#pragma config HPOL = ON      // Motor Control PWM High Side Polarity bit (PWM module high side output pins have active-high output polarity)
#pragma config PWMPIN = ON    // Motor Control PWM Module Pin Mode bit (PWM module pins controlled by PORT register at device Reset)

// FICD
#pragma config ICS = PGD1   // Comm Channel Select (Communicate on PGC1/EMUC1 and PGD1/EMUD1)
#pragma config JTAGEN = OFF // JTAG Port Enable (JTAG is Disabled)

```

</details>
  
----

[^1]: Image from [Microchip](https://www.microchip.com/en-us/product/dsPIC33FJ128MC802)'s [dsPIC33FJ128MC802 datasheet](https://ww1.microchip.com/downloads/aemDocuments/documents/OTH/ProductDocuments/DataSheets/70291G.pdf)


src/parameters.h
// parameters defines

//?  correct ?
#define SYS_FREQ 80000000L
#define FCY (SYS_FREQ/2)

// for filter.c :
// <editor-fold defaultstate="collapsed" desc="filter.c parameters">

#define FLOORS 4
// const short FLOOR_ORDER = 2;

#define THRESHOLD 230

// filter constants
#define M_FILTER 18 // Q18.14
#define FACTOR_FILTER 262144 // 1 << M
//#define FLOAT_FACTOR_FILTER (float) (factor) // doesn't work without parentheses because of <<'s priority, lost 10min just on this
#define FS_FILTER 15000 //? used ?

// </editor-fold>

// for motors.c :
// <editor-fold defaultstate="collapsed" desc="motors.c parameters">

#ifndef WHEELBASE
#define WHEELBASE 0.135
// in m
#endif

#ifndef WHEEL_RADIUS
#define WHEEL_RADIUS 0.04
// in m
#endif

#ifndef NUM_TICKS_PER_TURN
#define NUM_TICKS_PER_TURN 1430
// more or less
#endif

// PWM parameters and macro
#ifndef PWM_TIME_PERIOD
#define PWM_TIME_PERIOD 0.001
#endif
// In s, one period is one up AND one down,
// not just one 'hill' up
#ifndef CLOCK_PRESCALER_BINARY
#define CLOCK_PRESCALER_BINARY 0b00
#endif
// In ms
#ifndef CLOCK_PRESCALER
#define CLOCK_PRESCALER 1
#endif
// 0b00 corresponds to a scale of 1

// Max PWM value for the motors
#define MAX_PWM 0.8

#define QE1_CHANNEL_A_PIN 10 // Assigns channel A of QE1 to RP10(=RB10)
#define QE1_CHANNEL_B_PIN 11 // Assigns channel B of QE1 to RP11(=RB11)
#define QE2_CHANNEL_A_PIN 8  // Assigns channel A of QE2 to RP8(=RB8)
#define QE2_CHANNEL_B_PIN 9  // Assigns channel B of QE2 to RP9(=RB9)
#define CRUISE_ROTATION_SPEED 3.5   // rad / second
#define MAX_ACCELERATION_TIME_ROTATION 1.5 // seconds
#define MAX_ANGLE (CRUISE_ROTATION_SPEED*0.5*(MAX_ACCELERATION_TIME_ROTATION*2))
#define ACCELERATION_ROTATION (CRUISE_ROTATION_SPEED/MAX_ACCELERATION_TIME_ROTATION)
// name these defines in all caps
//#define acceptable_error_translation 0.04
//#define acceptable_error_rotation 0.06981 // = (float)4 / 360 * 2 * PI
#define ACCEPTABLE_ERROR_TRANSLATION 0.04
#define ACCEPTABLE_ERROR_ROTATION 0.06981 // = (float)4 / 360 * 2 * PI
//#define error_translation acceptable_error_transla+1.0
//#define error_rotation acceptable_error_rota+1.0
// change these to uppercase
//#define k_translation 3.69
//#define k_rotation (1.03*3)
#define K_TRANSLATION 3.69
#define K_ROTATION (1.03 * 3)
//#define time_interval 3 // ms
#define M1_DIR_PIN_AS_INPUT TRISBbits.TRISB2 // set to 0 to output
#define M2_DIR_PIN_AS_INPUT TRISBbits.TRISB3 // set to 0 to output

#define MOTOR_1_DIR_PIN LATBbits.LATB2
#define MOTOR_2_DIR_PIN LATBbits.LATB3
//#define maybeMOTOR_1_REVERSE LATBbits.LATB2
//#define maybeMOTOR_2_REVERSE LATBbits.LATB3

#define ACCELERATION 0.5 // seconds //?

// </editor-fold>

// <editor-fold defaultstate="collapsed" desc="FSM parameters">
#define FRAME_LENGTH 13 // Frame length in bits
#define FRAME_DATA_LENGTH 10 // Frame data length in bits
//#define FRAME_PARAMS_LENGTH FRAME_DATA_LENGTH-2 // Frame parameters length in bits
// </editor-fold>
